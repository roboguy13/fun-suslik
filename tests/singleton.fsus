%generate singleton[Int] Sll

data List := Nil | Cons Int List;

Sll : List >-> layout[x];
Sll (Nil) := emp;
Sll (Cons head tail) := x :-> head, (x+1) :-> tail, Sll tail;

singleton : Int -> List;
singleton x := Cons x (Nil);

