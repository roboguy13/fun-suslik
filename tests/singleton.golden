// *** Layout predicates ***

predicate ro_Sll(loc x) {
| (x == 0) => { emp }
| (not (x == 0)) => { [x,2] ** x :=> head ** (x+1) :=> tail ** ro_Sll(tail) }
}


// *** Function predicates ***

predicate singleton__ro_Sll__Int(int __p_0, loc __r_x) {
| true => { __r_x :-> __p_0 ** (__r_x+1) :-> 0 ** [__r_x,2] }
}


// *** Function specifications ***


void singleton(loc x1, loc r)
  { r :-> initialVal }
  { singleton__ro_Sll__Int(x1, r0) ** r :-> r0 }
{ ?? }

{r :-> initialVal}
{r :-> r0 ** singleton__ro_Sll__Int(x1, r0)<_alpha_1>}
void singleton (loc x1, loc r) {
  let r0 = malloc(2);
  *r = r0;
  *r0 = x1;
  *(r0 + 1) = 0;
}


--- Source AST size: 47

--- SuSLik AST size: 78
Succeeded
